event_callback.go

```diff
@@ -0,0 +1,239 @@
+package firego
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/zabawaba99/firego/sync"
+)
+
+// ChildEventFunc is the type of function that is called for every
+// new child added under a firebase reference. The snapshot argument
+// contains the data that was added. The previousChildKey argument
+// contains the key of the previous child that this function was called for.
+type ChildEventFunc func(snapshot DataSnapshot, previousChildKey string)
+
+// ChildAdded listens on the firebase instance and executes the callback
+// for every child that is added.
+func (fb *Firebase) ChildAdded(fn ChildEventFunc) error {
+	return fb.addEventFunc(fn, childAdded)
+}
+
+func childAdded(fn ChildEventFunc, notifications chan Event) {
+	var pk string
+	db := sync.NewDB()
+
+	orderAndSend := func(data map[string]interface{}) {
+		// we've got children so send an event per child
+		orderedKeys := make([]string, len(data))
+		var i int
+		for k := range data {
+			orderedKeys[i] = k
+			i++
+		}
+
+		sort.Strings(orderedKeys)
+
+		for _, k := range orderedKeys {
+			v := data[k]
+			node := sync.NewNode(k, v)
+			db.Add(k, node)
+			fn(newSnapshot(node), pk)
+			pk = k
+		}
+	}
+
+	for event := range notifications {
+		if event.Type != EventTypePut {
+			continue
+		}
+
+		child := strings.Split(event.Path[1:], "/")[0]
+		if event.Data == nil {
+			db.Del(child)
+			continue
+		}
+
+		if _, ok := db.Get("").Child(child); ok {
+			// if the child isn't being added, forget it
+			continue
+		}
+
+		m, ok := event.Data.(map[string]interface{})
+		if child == "" && ok {
+			// if we were given a map at the root then we have
+			// to send an event per child
+			orderAndSend(m)
+			continue
+		}
+
+		// we have a single event to process
+		node := sync.NewNode(child, event.Data)
```

Comment:
What do you think about L72-76 and L40-43 being extracted into a function that handles this?